### ISSUE TITLE
Sub-second resolution for timestamps lost on load

### ISSUE THREAD

**Issue by mwho** [databricks/spark-redshift#214]
I have a redshift table with a timestamp column (t_stamp) containing times at sub second resolution, i.e.

2016-05-27 00:27:41.0
2016-05-27 00:27:41.1
2016-05-27 00:27:41.2
2016-05-27 00:27:41.3

The times are of data type "timestamp" in redshift.

I'm using spark-redshift to create a dataframe from the redshift table using 

val df  = sqlContext.read
    .format("com.databricks.spark.redshift")....

and the datafame is loaded into df. The t_stamp column has type "string" now.

It seems as if the sub-second time information in t_stamp is lost and every time stamp is "floored" to the nearest second, i.e.

2016-05-27 00:27:41.0
2016-05-27 00:27:41.0
2016-05-27 00:27:41.0
2016-05-27 00:27:41.0

Is this expected behavior for spark-redshift? It is important for us to have sub-second time resolution for our application.

**Comment by JoshRosen** [comment]
Which versions of Spark and `spark-redshift` are you using?

**Comment by mwho** [comment]
Spark: 1.6.1
`spark-redshift`: v0.6.0

I am using Databricks, if that matters.

**Comment by JoshRosen** [comment]
Yep, this is definitely a bug! I've spotted the issue and will open a PR to fix it.

**Comment by JoshRosen** [comment]
In a nutshell, Redshift is omitting trailing zeros when formatting timestamps as strings at 10th or 100th of a second of resolution. For example, the timestamp `2016-05-27 00:27:41.100` is output as `2016-05-27 00:27:41.1` by Redshift, which is less than the full three-digits of padding that we expect, causing `spark-redshift` to switch to parsing with a non-subsecond-resolution time format.

Making matters slightly worse, the Java SimpleDateFormat parser behaves counterintuitively with sub-second timestamps:

```scala
new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").parse("1970-01-01 00:00:00.1").getTime()
```

will return `1` rather than `100`.

Also, 

```scala
println(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.S").parse("1970-01-01 00:00:00.1000000").getTime())
println(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.S").parse("1970-01-01 00:00:00.1").getTime())
```

will print

```
1
1000000
```

In a nutshell when parsing it looks looks like the `S` format specifier says "take any number of digits and treat that as a number in milliseconds".

Other folks have hit this counterintuitive behavior as well: https://stackoverflow.com/questions/7159895/simpledateformat-is-not-parsing-the-milliseconds-correctly

It looks like one solution is to use `joda-time`, but this adds an extra dependency.
